src/connector/src/sink/kafka.rs

```diff
@@ -333,6 +333,8 @@ pub struct KafkaSinkWriter {
     schema: Schema,
     pk_indices: Vec<usize>,
     is_append_only: bool,
+    /// The buffer that stores <event_key_object, event_object>
+    future_record_buffer: Vec<(Option<Value>, Option<Value>)>,
```

Comment:
Here we use a vec as queue to buffer the unsent data. However, the implementation of rdkafka already has a queue, and therefore I don't think we need to maintain the unsent data here. Instead, we can send every row to kafka and buffer the returned future. On commit, we then join all the buffered future to wait to ensure all previously sent data are delivered.

Reply:
However, such approach will easily cause data pipe up in the send queue, because there is no back-pressure on the send_result of the kafka client. Therefore, we can maintain the total size of sent but undelivered data. We can set a max limit on the total size. On an attempt to send new data, if the total size is going to exceed the limit, we should wait for the delivery of previously sent data to release the quota for sending new data.

Reply:
My original refactor is to maintain a `future_record_buffer: Vec<FutureRecord<'a, K, P>>`, which represents the sent but undelivered data in the `FutureProducer`, but this requires the explicit lifetime declaration. After trying the type exercise for some time, I changed back to the current design since the `await` for `FutureRecord` is sort of "atomic".

Reply:
The trait bound is as below (take `send_result_inner` as an example).
```rust
async fn send_result_inner<'a, K, P>(
        &'a self,
        record: FutureRecord<'a, K, P>,
    ) -> core::result::Result<DeliveryFuture, (KafkaError, FutureRecord<'a, K, P>)>
    where
        K: ToBytes + ?Sized,
        P: ToBytes + ?Sized,
{
    // Actual code...
}
```

Reply:
I see, thanks for the clarification🥳!

> We can set a max limit on the total size

Any suggestion for this parameter🤔️?

Reply:
At the beginning it can be a parameter passed from outside instead of a fixed one.